src/interfaces/txmempool.h

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INTERFACES_TXMEMPOOL_H
+#define BITCOIN_INTERFACES_TXMEMPOOL_H
+
+#include <primitives/transaction.h> // For CTransactionRef
+#include <sync.h>                   // For RecursiveMutex
+#include <util/check.h>             // For Assert
+
+#include <memory>
+#include <vector>
+
+class CTxMemPool;
+class CCoinsViewCache;
+
+namespace interfaces {
+
+//! Interface for accessing a txmempool.
+class TxMempool
+{
+public:
+    explicit TxMempool(RecursiveMutex* mutex)
+        : m_mutex{mutex} {}
+    virtual ~TxMempool() {}
+
+    //! A transaction was updated
+    virtual void transactionUpdated() = 0;
+
+    //! Run consistency check
+    virtual void check(const CCoinsViewCache& coins) = 0;
+
+    //! Remove txs after block has been connected
+    virtual void removeForBlock(const std::vector<CTransactionRef>& vtx, unsigned int block_height) = 0;
+
+    //! Alias for the txmempool mutex or nullptr if no txmempool was passed in
+    RecursiveMutex* const m_mutex;
+
+    //! Return pointer to internal txmempool class
+    virtual CTxMemPool* txmempool() { return nullptr; }
+};
+
+class NoTxMempool final : public TxMempool
+{
+public:
+    explicit NoTxMempool()
+        : TxMempool{nullptr} {}
+
+    void transactionUpdated() override { invalidCall(); }
+    void check(const CCoinsViewCache& coins) override { invalidCall(); }
+    void removeForBlock(const std::vector<CTransactionRef>& vtx, unsigned int block_height) override { invalidCall(); }
+
+private:
+    void invalidCall() { Assert(false); }
```

Comment:
face16170176c83d04c68454c91eaf42c80e6ec7
nit: custom `Assert` has no benefits in this case
```suggestion
    void invalidCall() { assert(false); }
```

Reply:
I prefer `Assert` over `assert` because it comes with the following benefits:

* Can not accidentally be disabled via NDEBUG
* No need to update the code if more features get added to `Assert`, which are not part of `assert`

Reply:
Probably, it is worse to update [Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md) :)